internal/cmd/local/local/java_logs.go

```diff
@@ -0,0 +1,80 @@
+package local
+
+import (
+	"bufio"
+	"io"
+	"regexp"
+	"strings"
+)
+
+// 2024-09-10 20:16:24 WARN i.m.s.r.u.Loggers$Slf4JLogger(warn):299 - [273....
+var javaLogRx = regexp.MustCompile(`^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} \x1b\[(?:1;)?\d+m(?P<level>[A-Z]+)\x1b\[m (?P<msg>\S+ - .*)`)
+
+type javaLogLine struct {
+	msg   string
+	level string
+}
+
+type javaLogScanner struct {
+	scanner *bufio.Scanner
+	line    javaLogLine
+}
+
+func newJavaLogScanner(r io.Reader) *javaLogScanner {
+	return &javaLogScanner{
+		scanner: bufio.NewScanner(r),
+		line: javaLogLine{
+			msg:   "",
+			level: "DEBUG",
+		},
+	}
+}
+
+func (j *javaLogScanner) Scan() bool {
+	for {
+		if ok := j.scanner.Scan(); !ok {
+			return false
+		}
+
+		// skip java stacktrace noise
+		if strings.HasPrefix(j.scanner.Text(), "\tat ") || strings.HasPrefix(j.scanner.Text(), "\t... ") {
+			continue
+		}
+
+		m := javaLogRx.FindSubmatch(j.scanner.Bytes())
+
+		if m != nil {
+			j.line.msg = string(m[2])
```

Comment:
I'm guessing `[0]` os the input string or the bytes leading up to the first capture group?

Reply:
> the text of the leftmost match of the regular expression in b

according to https://pkg.go.dev/regexp#Regexp.FindSubmatch